### PR TITLE
feat(spring-sdk): add support for event sourced entities

### PR DESCRIPTION
**Pull request by efgpinto**
Refs https://github.com/lightbend/kalix/issues/7360

Validations in place for event handler methods:
- they must be annotated with `@EventHandler`, be public, have a single parameter and a return type that matched the enclosing class type parameter
- there cannot be two handlers with the same type parameter (as in, each event type must have only one handler)

### CODE REVIEW

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/eventsourcedentity/EventSourcedEntitiesImpl.scala [hunk 1]
```diff
@@ -51,7 +51,7 @@ final class EventSourcedEntityService(
     val factory: EventSourcedEntityFactory,
     override val descriptor: Descriptors.ServiceDescriptor,
     override val additionalDescriptors: Array[Descriptors.FileDescriptor],
-    val anySupport: AnySupport,
+    val anySupport: MessageCodec,
```
Comment (octonato) [inline sdk/java-sdk/src/main/scala/kalix/javasdk/impl/eventsourcedentity/EventSourcedEntitiesImpl.scala:54]: `s/anySupport/messageCodec/`

I don't know if we did for the others. If not we can add in this PR as well.

**Review by octonato — COMMENTED** [review]
Looking good. 

For the integration tests, we have a sample to test wiring. Maybe we could simply extend the scope to add the ES. We don't need to add yet another sample.

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/eventsourcedentity/EventSourcedEntitiesImpl.scala (hunk 1 above)
Comment (efgpinto) [reply sdk/java-sdk/src/main/scala/kalix/javasdk/impl/eventsourcedentity/EventSourcedEntitiesImpl.scala:54]: yeah, will do that.. 👍

**Comment by efgpinto** [comment]
> For the integration tests, we have a sample to test wiring. Maybe we could simply extend the scope to add the ES. We don't need to add yet another sample.

Sounds good, will use the existing one instead. 👍

File: sdk/spring-sdk/src/main/java/kalix/springsdk/eventsourced/ReflectiveEventSourcedEntityProvider.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    Entity annotation = entityClass.getAnnotation(Entity.class);
+    if (annotation == null)
+      throw new IllegalArgumentException(
+          "Event Sourced Entity [" + entityClass.getName() + "] is missing '@Entity' annotation");
+
+    this.eventHandlers = EventSourcedHandlersExtractor.handlersFrom(entityClass);
+    if (this.eventHandlers.errors().nonEmpty()) {
+      throw new IllegalArgumentException(
+              "Event Sourced Entity [" + entityClass.getName() + "] has event handlers configured incorrectly: " + this.eventHandlers.errors());
```
Comment (efgpinto) [inline sdk/spring-sdk/src/main/java/kalix/springsdk/eventsourced/ReflectiveEventSourcedEntityProvider.java:67]: we validate every single possible handler and report all the errors found as opposed to stopping and throwing on first error. Example of what a log looks like:

```json
{"timestamp":"2022-09-02T09:23:23.467Z","thread":"main","logger":"org.springframework.boot.web.reactive.context.AnnotationConfigReactiveWebServerApplicationContext","message":"Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'kalixServer' defined in kalix.springsdk.KalixConfiguration: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [kalix.springsdk.impl.KalixServer]: Factory method 'kalixServer' threw exception; nested exception is java.lang.IllegalArgumentException: Event Sourced Entity [wiring.eventsourcedentity.Counter] has event handlers configured incorrectly: List(ValidationError(reason='must be public, with exactly one parameter and return type 'java.lang.Integer'', offendingMethods=List(notGoodBecausePrivate, wrongReturnType), ValidationError(reason='cannot have duplicate event handlers for the same event type: 'wiring.eventsourcedentity.ValueMultiplied'', offendingMethods=List(handleMultiply, duplicatedEvent))","context":"default","severity":"WARN"}
```

File: sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        public Integer publicMethodSimilarSignature(Integer event) {
+            return 0;
+        }
+
+        private Integer privateMethodSimilarSignature(Integer event) {
+            return 0;
+        }
+    }
```
Comment (efgpinto) [inline sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java:45]: Highlighting this one as an example of the syntax in use for annotating a class and its event handlers

**Review by efgpinto — COMMENTED** [review]
Opening up for review.

Validations in place for event handler methods:
- they must be annotated with `@EventHandler`, be public, have a single parameter and a return type that matched the enclosing class type parameter
- there cannot be two handlers with the same type parameter (as in, each event type must have only one handler)

File: sdk/java-sdk/src/main/scala/kalix/javasdk/impl/eventsourcedentity/EventSourcedEntitiesImpl.scala (hunk 1 above)
Comment (efgpinto) [reply sdk/java-sdk/src/main/scala/kalix/javasdk/impl/eventsourcedentity/EventSourcedEntitiesImpl.scala]: solved now

File: sdk/spring-sdk/src/it/java/com/example/wiring/SpringSdkWiringIntegrationTest.java [hunk 4]
```diff
@@ -74,4 +74,33 @@ public void verifyStreamActions() {
...
+            .post()
+            .uri("/counter/hello/multiply/20")
+            .retrieve()
+            .bodyToMono(String.class)
+            .block(timeout);
+
+    Assertions.assertEquals("\"200\"", counterMultiply);
+
+    String counterGet =
+        webClient.get().uri("/counter/hello").retrieve().bodyToMono(String.class).block(timeout);
```
Comment (franciscolopezsancho) [inline sdk/spring-sdk/src/it/java/com/example/wiring/SpringSdkWiringIntegrationTest.java:102]: How can a user know where the id of the entity is supposed to be passed? Like here `hello`

File: sdk/java-sdk/src/main/java/kalix/javasdk/eventsourcedentity/EventSourcedEntity.java [hunk 5]
```diff
@@ -82,7 +82,7 @@ public void _internalSetEventContext(Optional<EventContext> context) {
   /** INTERNAL API */
   public void _internalSetCurrentState(S state) {
     handlingCommands = true;
-    currentState = Optional.of(state);
```
Comment (He-Pin) [inline sdk/java-sdk/src/main/java/kalix/javasdk/eventsourcedentity/EventSourcedEntity.java]: NPE :)

File: sdk/spring-sdk/src/it/java/com/example/wiring/SpringSdkWiringIntegrationTest.java (hunk 4 above)
Comment (efgpinto) [reply sdk/spring-sdk/src/it/java/com/example/wiring/SpringSdkWiringIntegrationTest.java:102]: It's defined on the `RequestMapping` of the Entity but I'm pretty sure you know that.. :D so I'm not sure I understood the question.. Could you elaborate please? 🙇

File: sdk/java-sdk/src/main/java/kalix/javasdk/eventsourcedentity/EventSourcedEntity.java (hunk 5 above)
Comment (efgpinto) [reply sdk/java-sdk/src/main/java/kalix/javasdk/eventsourcedentity/EventSourcedEntity.java]: Exactly :)

File: sdk/spring-sdk/src/it/java/com/example/wiring/SpringSdkWiringIntegrationTest.java (hunk 4 above)
Comment (franciscolopezsancho) [reply sdk/spring-sdk/src/it/java/com/example/wiring/SpringSdkWiringIntegrationTest.java:102]: :)) No sir. I know little about Spring and totally missed it so I bow back to you :)

Comment (efgpinto) [reply sdk/spring-sdk/src/it/java/com/example/wiring/SpringSdkWiringIntegrationTest.java:102]: oh fair.. yeah, it definitely feels weird if you haven't seen it before.. cool :)

File: samples/spring-wiring-tests/src/main/java/wiring/eventsourcedentity/ValueIncreased.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class ValueIncreased {
+
+    public final int value;
+
+    @JsonCreator
+    public ValueIncreased(@JsonProperty("value") int value) {
```
Comment (octonato) [inline samples/spring-wiring-tests/src/main/java/wiring/eventsourcedentity/ValueIncreased.java:11]:
```suggestion
    public ValueIncreased(@JsonProperty int value) {
```

File: samples/spring-wiring-tests/src/main/java/wiring/eventsourcedentity/ValueMultiplied.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package wiring.eventsourcedentity;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class ValueMultiplied {
+    public final int value;
+
+    @JsonCreator
+    public ValueMultiplied(@JsonProperty Integer value) {
```
Comment (octonato) [inline samples/spring-wiring-tests/src/main/java/wiring/eventsourcedentity/ValueMultiplied.java:10]:
```suggestion
    public ValueMultiplied(@JsonProperty int value) {
```

to align with `ValueIncreased`

File: sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/eventsourcedentity/ReflectiveEventSourcedEntityRouter.scala [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import kalix.javasdk.eventsourcedentity.{ CommandContext, EventSourcedEntity }
+import kalix.javasdk.impl.eventsourcedentity.EventSourcedEntityRouter
+import kalix.springsdk.impl.{ ComponentMethod, InvocationContext }
+
+import java.lang.reflect.Method
+
+class ReflectiveEventSourcedEntityRouter[S, E <: EventSourcedEntity[S]](
+    override protected val entity: E,
+    componentMethods: Map[String, ComponentMethod],
+    handlerMethods: Map[Class[_], Method])
```
Comment (octonato) [inline sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/eventsourcedentity/ReflectiveEventSourcedEntityRouter.scala:29]: Maybe better if we call them `commandHandlerMethods` and `eventHandlerMethods`. 

Up to now we only had examples of `ComponentMethod`, but in reality they were all command handlers. 

Anyway, not that important. We can change this on another PR eventually.

**Review by octonato — APPROVED** [review]
LGTM

Left a few comments, but nothing essential. We can merge it and tackle small things later. It's up to you. 

Maybe worth to squash and merge this one.

**Pull request merged**